File: atcoder/abc/abc362/d/main.py
N, M = map(int, input().split())
A = [int(i) for i in input().split()]
G = [[] for _ in range(N)]
for i in range(M):
	u, v, b = map(int, input().split())
	u -= 1
	v -= 1
	G[u].append((v, b))
	G[v].append((u, b))

from heapq import heappush, heappop
inf = 10**18

def dijkstra(s, G, node=None):#s = start, n = number of the node
	if node==None: node = len(G)
	cost = [inf]*node
	hq = [(0, s)]
	seen = [False]*node
	cost[s] = A[s]
	while hq:
		d, v = heappop(hq)
		seen[v] = True
		if d > cost[v]: continue
		for to, weight in G[v]:
			if not seen[to] and cost[v] + weight + A[to] < cost[to]:
				cost[to] = cost[v] + weight + A[to]
				heappush(hq, (cost[to], to))
	return cost


ans = dijkstra(0, G)
print(*ans[1:])

File: atcoder/abc/abc362/d/main.cpp
#include <bits/stdc++.h>
using namespace std;
#define ll long long
const ll INF = 1e18;

vector<ll> dijkstra(int s, vector<vector<pair<int, ll>>> &G, vector<ll> &A){
  int node = G.size();
  vector<ll> cost(node, INF);
  priority_queue<pair<ll, int>, vector<pair<ll, int>>, greater<pair<ll, int>>> hq;
  vector<bool> seen(N, false);
  cost[s] = A[s];
  hq.push({0, s});
  while(!hq.empty()){
    int v = hq.top().second();
    seen[v] = true;
    while hq:
    v = 
  }
}

int main(){
  int N, M;
  cin >> N >> M;
  vector<ll> A(N);
  for(int i=0;i<N;i++){
    cin >> A[i];
  }
  vector<vector<pair<int, ll>>> G(N, vector<pair<int, ll>>());
  int u, v;
  ll b;
  for(int i=0;i<M;i++){
    cin >> u >> v >> b;
    u--;
    v--;
    G[u].push_back(make_pair(v, b));
    G[v].push_back(make_pair(u, b));
  }
}